Split into a library and a binary

sulix opened the pull request:
Some basic refactoring to split unciv into a library, which can be used by other code, and a binary which relies on it.

The code for writing files (including the png files for RIM images) is now a part of the binary, the libary technically doesn't depend on png (though Cargo.toml hasn't been updated to mention this).

The API is not perfect, but the binary should give a basic example on how to use it. ZfsEntry structs refer to an entry before it is read: you need to provide a reader to the same file to get its data. On the other hand, RimImage _does_ contain the data, once you load it from a ZfsEntry, there's no need to keep the file/reader around.

This is step one for getting the code into crates.io (#1) — feedback on the API is greatly desired. Also, there are several documentation cleanups, etc, which will happen once we've got an API which is not terrible to use.

sulix committed:
Split into a library and a binary

Some basic refactoring to split unciv into a library, which can be used
by other code, and a binary which relies on it.

The code for writing files (including the png files for RIM images) is now
a part of the binary, the libary technically doesn't depend on png (though
Cargo.toml hasn't been updated to mention this).

The API is not perfect, but the binary should give a basic example on how
to use it. ZfsEntry structs refer to an entry before it is read: you need
to provide a reader to the same file to get its data. On the other hand,
RimImage _does_ contain the data, once you load it from a ZfsEntry, there's
no need to keep the file/reader around.

justinfarrelldev commented:
This looks great. I especially appreciate the doc comments over the various functions going in-depth, I can't wait to use this in the client (it should be extremely ergonomic for what I currently have, and if I see an issue later on I'll definitely either put in a PR or make a new issue).

sulix committed:
Remove byteorder dependency

Use the binary_io module (which I've been using for my other projects) instead
of the byteorder crate. This lets us drop an external dependency (yay!), though
png requires it anyway, so it doesn't actually help as much as I'd've hoped.

Still, this should be an improvement!

sulix committed:
Make the png dependency optional.

Currently, we use the 'png' library to convert 'rim' images to png files.
However, this is constrained to the unciv program, and the library has no need
of the 'png' dependency.

Make the 'png' dependency optional so that users of the library don't gain a
new, heavy dependency. It's behind the 'png' feature, which is currently
disabled by default (though this may change… I'm still thinking about it!)

Note that 'png' is still at version 0.15 (to match the original's support of
Debian bullseye). Additionally, note that the 'png' dependency is not an
implicit dependency (though that might be better) as they're not supported by
cargo 1.46. That, too, may change at some point.

If 'png' is disabled, unciv will extract .rim files as they are.

sulix committed:
README: Update for the conversion into a library.

There are a bunch of small changes as a part of making unciv usable as a
library. Hopefully this update to the documentation covers them.
